python/ray/serve/backend_state_manager.py

```diff
@@ -948,23 +1039,55 @@ def _completed_goals(self) -> List[GoalId]:
+            target_version = self._target_versions[backend_tag]
+
+            all_running_replica = self._replicas[backend_tag].get(
                 states=[ReplicaState.RUNNING])
-            running_count = len(running)
+            running_at_target_version_replica = [
+                replica.version == target_version
+                for replica in all_running_replica
+            ]
+            failed_to_start_replica = self._replicas[backend_tag].get(
+                states=[ReplicaState.FAILED_TO_START])
 
             # Check for deleting.
-            if target_count == 0 and running_count == 0:
+            if target_replica_count == 0 and len(all_running_replica) == 0:
                 deleted_backends.append(backend_tag)
                 completed_goals.append(
                     self._backend_goals.pop(backend_tag, None))
 
             # Check for a non-zero number of backends.
-            elif target_count == running_count:
-                # Check that all running replicas are the target version.
-                target_version = self._target_versions[backend_tag]
-                if all(r.version == target_version for r in running):
-                    completed_goals.append(
-                        self._backend_goals.pop(backend_tag, None))
+            elif target_replica_count == len(
+                    running_at_target_version_replica):
+                completed_goals.append(
+                    self._backend_goals.pop(backend_tag, None))
+
+            elif len(failed_to_start_replica) > 0 and len(
+                    running_at_target_version_replica) > 0:
+                # Reconfigure action is considered partially successful since
+                # we had at least one running replica at target version, thus
+                # let user code return but serve controller will keep trying
+                # to bring to target running replica fully
+                completed_goals.append(
+                    self._backend_goals.pop(backend_tag, None))
+                # Clear failed to start replicas as the deployment is
+                # considered done with partially fulfilled target replicas
+                self._replicas[backend_tag].pop(
+                    states=[ReplicaState.FAILED_TO_START])
+
+            elif len(failed_to_start_replica) == target_replica_count:
+                # Reconfigure action failed as all replicas failed to start
+                # without any running instance.
+                failed_to_start_goals.append(
+                    self._backend_goals.pop(backend_tag, None))
+                # Clear failed to start replicas as the deployment is
+                # considered failed completely with exception set
+                self._replicas[backend_tag].pop(
+                    states=[ReplicaState.FAILED_TO_START])
+                # TODO: (jiaodong) Determine if we should delete backend or
+                # roll back to previous version
```

Comment:
What is the current behavior here? It looks like we'll just continue trying to deploy the new replicas forever, right? We definitely want to avoid that as it puts the controller effectively in a while True loop.